clients/python/airavata_custos/security/keycloak_connectors.py

```diff
@@ -0,0 +1,135 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import time
+from oauthlib.oauth2 import LegacyApplicationClient
+from requests_oauthlib import OAuth2Session
+import requests
+from airavata_custos import settings
+
+
+class KeycloakBackend(object):
```

Comment:
I think you'll need a method to initiate the redirect flow, unless you are planning on doing that in a separate PR.

Reply:
Not sure about what the redirect flow is. I have added 3 flows 1) user_authentication: when user name and password are given 2) account_authentication: when the service account client id and client credentials are supplied 3) refresh token: when user/account is already authenticated 
My understanding is the redirect flow is when clientId and client credentials are supplied ie option2 account_authentication. Please correct me if I am wrong.

Reply:
There are two parts to the redirect flow:
1. Redirect to the IdP
2. Process the response

You've implemented the second one. We could also have a utility code for generating the approach redirection URL for initiating the first part.

In the Django portal, setting up the redirect URL is handled here: https://github.com/apache/airavata-django-portal/blob/master/django_airavata/apps/auth/views.py#L42